Sources/Algorithms/Stride.swift

```diff
@@ -0,0 +1,160 @@
+// This source file is part of the Swift Algorithms open source project
+//
+// Copyright (c) 2020 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+//
+//===----------------------------------------------------------------------===//
+
+//===----------------------------------------------------------------------===//
+// striding(by:)
+//===----------------------------------------------------------------------===//
+
+extension Collection {
+  /// Returns a collection stepping through the elements every `step` starting
+  /// at the first value. Any remainders of the stride will be trimmed.
+  ///
+  ///     (0...10).striding(by: 2) // == [0, 2, 4, 6, 8, 10]
+  ///     (0...10).striding(by: 3) // == [0, 3, 6, 9]
+  ///
+  /// - Complexity: O(1). Access to successive values is O(1) if the
+  /// collection conforms to `RandomAccessCollection`; otherwise,
+  /// O(_k_), where _k_ is the striding `step`.
+  ///
+  /// - Parameter step: The amount to step with each iteration.
+  /// - Returns: Returns a collection stepping through the elements by the
+  /// specified amount.
+  public func striding(by step: Int) -> Stride<Self> {
+    Stride(base: self, stride: step)
+  }
+}
+
+public struct Stride<Base: Collection> {
+  
+  public let base: Base
+  public let stride: Int
+  
+  init(base: Base, stride: Int) {
+    precondition(stride > 0, "striding must be greater than zero")
+    self.base = base
+    self.stride = stride
+  }
+}
+
+extension Stride {
+  public func striding(by step: Int) -> Self {
+    Stride(base: base, stride: stride * step)
+  }
+}
+
+extension Stride: Collection {
+  
+  public struct Index: Comparable {
+    
+    let base: Base.Index
+    
+    init(_ base: Base.Index) {
+      self.base = base
+    }
+    
+    public static func < (lhs: Index, rhs: Index) -> Bool {
+      lhs.base < rhs.base
+    }
+  }
+  
+  public var startIndex: Index {
+    Index(base.startIndex)
+  }
+  
+  public var endIndex: Index {
+    Index(base.endIndex)
+  }
+  
+  public subscript(i: Index) -> Base.Element {
+    base[i.base]
+  }
+  
+  public func index(after i: Index) -> Index {
+    precondition(i.base < base.endIndex, "Advancing past end index")
+    return base.index(i.base, offsetBy: stride, limitedBy: base.endIndex)
+      .map(Index.init) ?? endIndex
+  }
+  
+  public func index(
+    _ i: Index,
+    offsetBy distance: Int,
+    limitedBy limit: Index
+  ) -> Index? {
+    base.index(i.base, offsetBy: distance * stride, limitedBy: limit.base)
+      .map(Index.init)
+  }
+  
+  public var count: Int {
+    (base.count - 1) / stride + 1
+  }
+  
+  public func distance(from start: Index, to end: Index) -> Int {
+    base.distance(from: start.base, to: end.base) / stride
+  }
+  
+  public func index(_ i: Index, offsetBy distance: Int) -> Index {
+    Index(base.index(i.base, offsetBy: distance * stride))
```

Comment:
Similar issue — this can overrun `base.endIndex` in a way that will trap in some collections when it should just return `endIndex`. I think this will have the problem:

```swift
let a = 1...5
let b = a.striding(by: 3) // equivalent to [1, 4]
let i = b.index(b.startIndex, offsetBy: 2)
```

Reply:
I was a little unsure about what to do here, when using the standard collection API for `offsetBy` it trapped as you say so I mimic'd that behaviour.

Is it better to phrase such that it returns `endIndex` like such:
```swift
  public func index(_ i: Index, offsetBy distance: Int) -> Index {
    precondition(i.base < base.endIndex, "Advancing past end index")
    let limit = distance > 0 ? endIndex : startIndex
    return index(i, offsetBy: distance, limitedBy: limit) ?? limit
  }
```